### DISCUSSION TITLE
[help!!!!]how can i get to know the messages (using the send(msgs, timeout) func) have been successfully acked on the CANBUS?

### DISCUSSION THREAD

**Discussion by leijr** [hardbyte/python-can#1822] [Q&A]
for example, i send msgs in a very high frequency, but i dont know if the msg i send successfully acked on the can bus.
i try to send next message after the previous message has been acked by the can bus
is there any functions or api can help me resolve this? 
 really appreciated!!! thank you very much

**Comment by Tbruno25** [comment]
What's the use case? The protocol itself handles ack's and you should never really need to verify this.

**Reply by leijr** [reply]
i have checked the user docuemnt:
there is a api function called bus.send(msg, tiimeout=None). timeout (Optional[float]) – If > 0, wait up to this many seconds for message to be ACK’ed or for transmit queue to be ready depending on driver implementation.
but my purpose is sending many messages in a very high frequency in a python for loop, then check if the last message is send successfully. 
**the Pseudo code  :**
**for i in range(1, len(send_data)): 
     msg = can.Message(arbitration_id=send_id, data=send_data[i], is_extended_id=False, dlc=len(send_data[i]), is_fd=is_fd)
     self.bus.send(msg, timeout = 100)
      # print("data", send_data[i])
36_service = [0x02, 0x36, 0xaa,.....]
self.bus.send(36_service)
uds_msg_response = self.bus.recv(timeout=0.01)**

the upper code after i debug i found that  i send too many message in the for loop , all the message are on the transmit queue to be  sent, if i continue send a new uds message service and wait the uds response, the  self.bus.recv will be timeout because of the vector are still sending the message on the transmit queque, and the response delayed and timeout
**so i want to to know if the last message has been acked by the can bus, i continue sending  the next new message by the following steps.**

**Reply by Tbruno25** [reply]
I think I understand your reasoning, but why not simply increase the timeout? 

If you don't want to do that for whatever reason, another option could be to utilize the `receive_own_messages` parameter during bus creation. I have no idea whether this will function the way you need but it's something to play around with. For example, you can verify you're able to receive the last message before transmitting the next --

```
import can

bus = can.Bus(channel="demo", interface="virtual", receive_own_messages=True)
message = can.Message(arbitration_id=0x123, data=[0x11, 0x22, 0x33, 0x44])

for _ in range(25):
    bus.send(message)
    
    while received := bus.recv():
        if received.arbitration_id == message.arbitration_id:
            if received.data == message.data:
                break
```
